kong/router/atc_compat.lua

```diff
@@ -0,0 +1,586 @@
+local ngx_WARN = ngx.WARN
+
+
+
+local SLASH         = byte("/")
+local MAX_HEADER_COUNT = 255
+local MAX_REQ_HEADERS = 100
+
+
+local function is_regex_magic(path)
+  return sub(path, 1, 1) == "~"
+end
+
+
+function _M._set_ngx(mock_ngx)
+  if type(mock_ngx) ~= "table" then
+    return
+  end
+
+  if mock_ngx.header then
+    header = mock_ngx.header
+  end
+
+  if mock_ngx.var then
+    var = mock_ngx.var
+  end
+
+  if mock_ngx.log then
+    ngx_log = mock_ngx.log
+  end
+
+  if type(mock_ngx.req) == "table" then
+    if mock_ngx.req.get_method then
+      get_method = mock_ngx.req.get_method
+    end
+
+    if mock_ngx.req.get_headers then
+      get_headers = mock_ngx.req.get_headers
+    end
+  end
+end
+
+
+local protocol_subsystem = constants.PROTOCOLS_WITH_SUBSYSTEM
+
+
+local function gen_for_field(name, op, vals, vals_transform)
+  local values_n = 0
+  local values = {}
+
+  if vals then
+    for _, p in ipairs(vals) do
+      values_n = values_n + 1
+      local op = (type(op) == "string") and op or op(p)
+      values[values_n] = name .. " " .. op ..
+                         " \"" .. (vals_transform and vals_transform(op, p) or p) .. "\""
+    end
+
+    if values_n > 0 then
+      return "(" .. tb_concat(values, " || ") .. ")"
+    end
+  end
+
+  return nil
+end
+
+
+local OP_EQUAL = "=="
+local OP_PREFIX = "^="
+local OP_POSTFIX = "=^"
+local OP_REGEX = "~"
+
+
+local function get_atc(route)
+  local out = {}
+
+  local gen = gen_for_field("net.protocol", OP_EQUAL, route.protocols)
+  if gen then
+    tb_insert(out, gen)
+  end
+
+  local gen = gen_for_field("http.method", OP_EQUAL, route.methods)
+  if gen then
+    tb_insert(out, gen)
+  end
+
+  local gen = gen_for_field("tls.sni", OP_EQUAL, route.snis)
+  if gen then
+    tb_insert(out, gen)
+  end
+
+  local gen = gen_for_field("http.host", function(host)
+    if host:sub(1, 1) == "*" then
+      -- postfix matching
+      return OP_POSTFIX
+    end
+
+    if host:sub(-1) == "*" then
+      -- prefix matching
+      return OP_PREFIX
+    end
+
+    return OP_EQUAL
+  end, route.hosts, function(op, p)
+    if op == OP_POSTFIX then
+      return p:sub(2)
+    end
+
+    if op == OP_PREFIX then
+      return p:sub(1, -2)
+    end
+
+    return p
+  end)
+  if gen then
+    tb_insert(out, gen)
+  end
+
+  local gen = gen_for_field("http.path", function(path)
+    return is_regex_magic(path) and OP_REGEX or OP_PREFIX
+  end, route.paths, function(op, p)
+    if op == OP_REGEX then
+      return sub(p, 2):gsub("\\", "\\\\")
+    end
+
+    return normalize(p, true)
+  end)
+  if gen then
+    tb_insert(out, gen)
+  end
+
+  if route.headers then
+    local headers = {}
+    for h, v in pairs(route.headers) do
+      local single_header = {}
+      for _, ind in ipairs(v) do
+        local name = "any(http.headers." .. h:gsub("-", "_"):lower() .. ")"
+        local value = ind
+        local op = OP_EQUAL
+        if ind:sub(1, 2) == "~*" then
```

Comment:
`~*` should be `~` or not?

Reply:
This is for header regex, and the operator is different.